_posts/2015-04-17-Enums-match-mutation-and-moves.md

```diff
@@ -0,0 +1,613 @@
+    assert_eq!(num_to_ordinal(   1),    "1st");
+    assert_eq!(num_to_ordinal(  12),   "12th");
+    assert_eq!(num_to_ordinal(  22),   "22nd");
+    assert_eq!(num_to_ordinal(  43),   "43rd");
+    assert_eq!(num_to_ordinal(  67),   "67th");
+    assert_eq!(num_to_ordinal(1901), "1901st");
+}
+```
+
+The Rust compiler accepts the above program; this is notable because
+its static analysis is ensuring both that `suffix` is always
+initialized before we run the `format!` at the end *and* that `suffix`
+is assigned at most once during the function's execution (because if
+we could assign `suffix` multiple times, the compiler would force us
+to mark `suffix` as mutable).
+
+To be clear, the above program certainly *can* be written in an
+expression-oriented style in Rust; for example, like so:
+
+```rust
+fn num_to_ordinal_expr(x: u32) -> String {
+    format!("{}{}", x, match (x % 10, x % 100) {
+        (1, 1) | (1, 21...91) => "st",
+        (2, 2) | (2, 22...92) => "nd",
+        (3, 3) | (3, 23...93) => "rd",
+        _                     => "th"
+    })
+}
+```
+
+Sometimes expression-oriented style can yield very succinct code;
+other times the style requires contortions that can be
+avoided by writing in a statement-oriented style.
+
+Each of the styles has its use cases. Crucially, switching to a
+statement-oriented style in Rust does not sacrifice every other
+feature that Rust provides, such as the guarantee that a non-`mut`
+binding is assigned at most once.
+
+An important case where this arises is when one wants to
+initialize some state and then borrow from it, but only on
+*some* control-flow branches.
+
+```rust
+fn sometimes_initialize(input: i32) {
+    let string: String; // a dynamically-constructed string value
+    let borrowed: &str; // a reference to string data
+    match input {
+        0...100 => {
+			// Construct a String on the fly...
+            string = format!("input prints as {}", input);
+			// ... and then borrow from inside it.
+            borrowed = &string[6..];
+        }
+        _ => {
+			// String literals are *already* borrowed references
+            borrowed = "expected between 0 and 100";
+        }
+    }
+    println!("borrowed: {}", borrowed);
+
+    // Below would cause compile-time error if uncommented...
+
+	// println!("string: {}", string);
+
+	// ...namely: error: use of possibly uninitialized variable: `string`
+}
+
+#[test]
+fn demo_sometimes_initialize() {
+    sometimes_initialize(23);  // this invocation will initialize `string`
+    sometimes_initialize(123); // this one will not
+}
+```
+
+The interesting thing about the above code is that after the `match`,
+we are not allowed to directly access `string`, because the compiler
+requires that the variable be initialized on every path through the
+program. At the same time, we *can*, via `borrowed`, access the data that
+is held *within* `string`, because a reference to that data is held by the
```

Comment:
"that *may* be held"?  (Or just "held"?) The text "is held" conveys stronger property than i wished